src/io.rs

```diff
@@ -0,0 +1,95 @@
+use std::io::{Error, Read, Result, Write};
+
+use libc::{c_void, size_t};
+
+use crate::error::rustls_io_error;
+
+/// A callback for rustls_server_session_read_tls or rustls_client_session_read_tls.
+/// An implementation of this callback should attempt to read up to n bytes from the
+/// network, storing them in `buf`. If any bytes were stored, the implementation should
+/// set out_n to the number of bytes stored and return 0. If there was an error,
+/// (including EAGAIN or EWOULDBLOCK), the implementation should return `errno`.
```

Comment:
`crustls` cares about `errno` semantics because it will marshal an `errno` value into a `Result<_, io::Error>`, and `io::Error` knows how to pass through errnos. But you make a good point - I could probably embed arbitrary data inside an `io::Error` in its wrapped error. Are you thinking "arbitrary platform int" as the return type?

Reply:
I think it has to be. Passing the error returned from the client callback through [from_raw_os_error](https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error) and then back into an int with `std::io::Error::raw_os_error` won't work reliably unless you can guarantee that the callback returned an errno that `std::io::Error` thinks is valid. I think most values will end up as `io::ErrorKind::Other`, but you can get more perniciously incorrect errors back if the range of errors the callback returns are not errnos but their values overlap with the range of valid errnos. I suspect the best you can do is preserve the integer value returned by the callback unmodified, and trust the client of `crustls` to decide if that value is an errno or if it is drawn from some other range of error codes.